tket2/src/portmatching/constraint.rs

```diff
@@ -0,0 +1,237 @@
+                    let out_port = out_port.as_outgoing().expect("wrong port type");
+
+                    // Get the incoming ports
+                    let exp_in_ports = vals[1..]
+                        .iter()
+                        .map(|&(n, p)| (n, p.as_incoming().expect("wrong port type")))
+                        .sorted();
+
+                    // Get the actual incoming ports in the hugr
+                    let actual_in_ports = hugr.linked_inputs(out_node, out_port).sorted();
+
+                    exp_in_ports.eq(actual_in_ports)
+                } else {
+                    // In this case we just need to check that the incoming
+                    // ports are all connected to the same wire.
+
+                    // nothing to do if there is just one port
+                    if vals.len() <= 1 {
+                        return true;
+                    }
+
+                    // The outgoing port of a wire is unique, so check that it
+                    // is the same for all the incoming ports.
+                    let mut first_out_port = None;
+                    for (node, port) in vals {
+                        let in_port = port.as_incoming().expect("wrong port type");
+                        let Some(out_port) = hugr.single_linked_output(node, in_port) else {
+                            return false;
+                        };
+                        if first_out_port.is_none() {
+                            first_out_port = Some(out_port);
+                        } else if out_port != first_out_port.unwrap() {
+                            return false;
+                        }
+                    }
+                    true
+                }
+            }
+            Predicate::IsNotEqual { .. } => {
+                // Get the native hugr node values
+                let vals = unwrap_to_hugr_values::<hugr::Node, _>(args);
+                let first_val = vals[0];
+                vals[1..].iter().all(|&v| v != first_val)
+            }
+        }
+    }
+}
+
+/// Convert a port to its unique outgoing port.
+///
+/// If the port is outgoing, leave it unchanged, otherwise find the unique
+/// outgoing port that it is attached to.
+fn to_out_port(
+    hugr: &impl HugrView,
+    (node, port): (hugr::Node, hugr::Port),
+) -> (hugr::Node, hugr::OutgoingPort) {
+    match port.as_directed() {
+        Either::Left(in_port) => hugr
+            .single_linked_output(node, in_port)
+            .expect("not a single outgoing port"),
+        Either::Right(out_port) => (node, out_port),
+    }
+}
+
+fn unwrap_to_hugr_values<'b, V, B>(args: impl IntoIterator<Item = &'b B>) -> Vec<V>
+where
+    B: Borrow<HugrVariableValue> + 'b,
+    V: TryFrom<HugrVariableValue>,
+    V::Error: Debug,
+{
+    args.into_iter()
+        .map(|arg| {
+            let var = arg.borrow().clone();
+            var.try_into().expect("invalid variable binding type")
```

Comment:
Unsure about the `expect` here, since `Predicate` is a public enum we should fail more gracefully on badly constructed elems.
Would it make sense to just return `false` from `pm::Predicate::check` if the `try_into` fails?

Reply:
I see. I still think a mistyped predicate should be a hard failure (I don't want them to sneak around silently). But you are probably right that I should amend `Predicate::check` to return a Result, and then unwrap this further up the stack, where it would be a proper logic error.